dali/pipeline/operators/geometric/bb_flip_test.cc

```diff
@@ -0,0 +1,176 @@
+ * 2. Upper-left and Lower-right corners
+ *    (x1, y1, x2, y2)
+ *
+ * Both of them have coordinates in image coordinate system
+ * (i.e. 0.0-1.0)
+ */
+struct Roi {
+  float roi[BB_STRUCT_SIZE];
+};
+
+/**
+ * Functor for calculating Roi hash
+ */
+struct RoiHash {
+  std::size_t operator()(Roi const &roi) const noexcept {
+    std::stringstream ss;
+    ss << std::to_string(roi.roi[0]) << std::to_string(roi.roi[1])
+       << std::to_string(roi.roi[2]) << std::to_string(roi.roi[3]);
+    return std::hash<std::string>{}(ss.str());
+  }
+};
+
+
+bool operator==(const Roi &lh, const Roi &rh) noexcept {
+  return RoiHash{}(lh) == RoiHash{}(rh);  //NOLINT
+}
+
+
+std::unordered_map<Roi, Roi, RoiHash> wh_rois = {
+        {{.2,  .2,  .4, .3}, {.4,  .2,  .4, .3}},
+        {{.0,  .0,  .5, .5}, {.5,  .0,  .5, .5}},
+        {{.3,  .2,  .1, .1}, {.6,  .2,  .1, .1}},
+        {{.0,  .0,  .2, .3}, {.8,  .0,  .2, .3}},
+        {{.0,  .0,  .1, .1}, {.9,  .0,  .1, .1}},
+        {{.5,  .5,  .1, .1}, {.0,  .5,  .1, .1}},
+        {{.0,  .6,  .7, .4}, {.3,  .6,  .7, .4}},
+        {{.6,  .2,  .3, .3}, {.1,  .2,  .3, .3}},
+        {{.4,  .3,  .5, .5}, {.1,  .3,  .5, .5}},
+        {{.25, .25, .5, .5}, {.25, .25, .5, .5}},
+};
+
+std::unordered_map<Roi, Roi, RoiHash> two_pt_rois = {
+        {{.2,  .2,  .6,  .5},  {.4,  .2,  .8,  .5}},
+        {{.0,  .0,  .5,  .5},  {.5,  .0,  1.,  .5}},
+        {{.3,  .2,  .4,  .3},  {.6,  .2,  .7,  .3}},
+        {{.0,  .0,  .2,  .3},  {.8,  .0,  1.,  .3}},
+        {{.0,  .0,  .1,  .1},  {.9,  .0,  1.,  .1}},
+        {{.5,  .5,  .6,  .6},  {.4,  .5,  .5,  .6}},
+        {{.0,  .6,  .7,  .9},  {.3,  .6,  1.,  .9}},
+        {{.6,  .2,  .3,  .3},  {.1,  .2,  .4,  .5}},
+        {{.4,  .3,  .9,  .8},  {.1,  .3,  .6,  .8}},
+        {{.25, .25, .75, .75}, {.25, .25, .75, .75}},
+};
+
+using RoiMap = std::unordered_map<Roi, Roi, RoiHash>;
+
+
+/**
+ * Flatten RoiMap, so that it is a vector of continuous floats
+ * @param keys if true, the continuous float will be obtained from map keys
+ *             if false - from map values
+ * @return flattened vector
+ */
+std::vector<float> flatten(const RoiMap &roi_map, bool keys) {
+  std::vector<float> ret;
+  for (const auto &it : roi_map) {
+    auto _it = keys ? it.first.roi : it.second.roi;
+    ret.insert(ret.end(), _it, _it + BB_STRUCT_SIZE);
+  }
+  return ret;
+}
+
+}  // namespace
+
+template<typename ImageType>
+class BbFlipTest : public DALISingleOpTest<ImageType> {
+ protected:
+  std::vector<TensorList<CPUBackend> *>
+  Reference(const std::vector<TensorList<CPUBackend> *> &inputs, DeviceWorkspace *ws) override {
+    TensorList<CPUBackend> batch;
+    batch.Resize(new_batch_size_);
+    auto *out_data = batch.mutable_data<float>();
+
+    auto rois = flatten(*test_data_, false);
+    std::memcpy(out_data, rois.data(), BB_STRUCT_SIZE * sizeof(float) * test_data_->size());
+
+    vector<TensorList<CPUBackend> *> ret(1);
+    ret[0] = new TensorList<CPUBackend>();
```

Comment:
Maybe just resize this tensor list that you created here and flatten data directly to it, instead of allocating 3 times.

Reply:
Right, that was just a bad example of copy-paste. I'll refactor it